### ISSUE TITLE
Deduplication of files added with wildcards

### ISSUE THREAD

**Issue by apawsey** [oleg-shilo/wixsharp#270]
Hi,

First obviously has to be said, thank you for this project.  Amazing compared to working with WIX directly!

One of the major reasons we wanted to use this was managing our files list with wildcards. The fact that this isn't easily accomplished with WIX itself boggles my mind, but WixSharp to the rescue!

However, I've now got a slightly obvious, but seems surely easy to resolve, issue.  I'm adding files based on the output folders of several different projects, in our case: a server, a client, and an excel plugin.  The excel plugin and client should be installed in the same folder, but they share some assemblies.  So now when I add like this:

```
ManagedProject project = new ManagedProject("MyProject",
                new Dir(@"%ProgramFiles%\MyProject",
                    new Dir("Client",
                        new DirFiles(clientFeature, $@"..\MyProject.Application\bin\{buildPath}\*.*"),
                        new DirFiles(excelPluginFeature, $@"..\MyProject.Excel\bin\{buildPath}\*.*")),
                    new Dir("Server",
                        new DirFiles(serverFeature, $@"..\MyProject.Service.Web\bin\{buildPath}\net461\win8-x86\*.*"))));
```
... I get errors from WIX that some files are being duplicated.

```
D:\Workspaces\MyProject\MyProject.Install\MyProject.wxs(363): error LGHT0204: ICE30: The target file 'h9s3
dwob.dll|EntityFramework.dll' is installed in '[ProgramFilesFolder]\MyProject\Client\' by two different components on an
LFN system: 'Component.EntityFramework.dll_789477219' and 'Component.EntityFramework.dll_1069612794'. This breaks compo
nent reference counting. [D:\Workspaces\MyProject\MyProject.Install\MyProject.Install.csproj]
```

I sort of thought it might deduplicate on it's own, but failing that, I thought surely I can write a linq query type thing to dedupe them.  However I then realised there is some potential issues on things like whether the file is the same size etc.  And I now realise that no information about the file, other than the name, is available during the WixSharp execution, so that might be complicated.

So I'm partly looking for your advice on how this should probably be solved...?
I'm thinking of submitting a pull request with a couple of changes along these lines...

- Add a property to File objects, something like OriginalFileInfo, where we make available source path, file size, and any other properties that might be relevant.
- Add extension methods like `DedupeFilesBy()` where we allow the specification of a comparision function, so compare by file name and then by size for example, and throw an error if we find things like duplicate names, but non duplicate sizes (which would indicate different versions of files).

Sorry, I need to think that through a little more, but you hopefully get the idea?  Thoughts?

**Comment by apawsey** [comment]
As a quick fix, I've added this to my project (using c# 7.0 local functions)...

```
project.ResolveWildCards();
void DedupeFolder(Dir dir)
{
    if (dir.Files.Any())
        dir.Files = dir.Files.DistinctBy(x => Path.GetFileName(x.Name)).ToArray();
    foreach (Dir childDir in dir.Dirs)
    {
        DedupeFolder(childDir);
    }
}

foreach (Dir projectDir in project.Dirs)
{
    DedupeFolder(projectDir);
}

```
Obviously like I said, this isn't exactly great because things like differing versions would not be caught...

**Comment by manytwo** [comment]
Hi, Call dedupe on "global scope" can cause strange behaviors.
Is this case should not be resolved by another "design" of your features and files?

Like Add feature with shared component, or if those components can't be added separetly, merge into one before msi build?

I faced this problem before ([#48](https://github.com/oleg-shilo/wixsharp/issues/48)). It is not a very common scenario I guess.

Think about this use case: 
```
src: BaseFeature\fileA.txt
src: ExtendedFeature\fileA.txt
```
BaseFeature and ExtendedFeature target same folder. Wix compiler is warning you and do not compile.
think about different use cases based on this configuration (same as original post):

Features (optional for MSI Technology) are not installed together:
One of them will have the file missing on target computer.

You want to replace base file if you install your second feature:
fileA.txt is from BaseFeature, or from ExtendedFeature if it is also installed (a case I was facing).

It is not really a good idea to "dedupe" files, they are in different features. 

You can disable ICE:30 validation 
```csharp
Compiler.LightOptions += " -sice:ICE30";
```
 or split your shared dll in a common feature.

This helper dedupe function should obviously be a good add, but only as helper to call manually of required dirs. Call it on each dir can easily be done by developers. 

And before use it, be aware of all uses cases that can happen with this configuration! This can be a trap (I falled into) when extending your installer. For now your two features are installed together.  For now.. :)

**Comment by oleg-shilo** [comment]
> Hi, Call dedupe on "global scope" can cause strange behaviors.

I think there is a slight confusion. It has to be on global scope for all directories resolved with wild card. Though this does not mean that _dedup_ needs to be performed automatically. 

The c48be6ea73a99fcc4dc5cd28d6f429bfee3ee68d commit delivered the dedup algorithm implemented. It's done via `project.WildCardDedup`. If you don't assign it then the WiX compiling approach is as normal. But if you want to control the way wild cards are resolved, then you can assign `project.WildCardDedup` to any routine of your choice. Just remember, `project.WildCardDedup` is called for every `Dir` created during wild card resolving: 

```C#
project.WildCardDedup = dir =>
{
    if (dir.Files.Any())
        dir.Files = dir.Files.DistinctBy(x => x.Name.PathGetFileName()).ToArray();
};
// or built-in routine
project.WildCardDedup = Project.UniqueFileNameDedup;
...
Compiler.BuildMsi(project);
```

**Comment by manytwo** [comment]
Ok nice, as usual you did it the right way :)
Users of
 ``` csharp
project.WildCardDedup = Project.UniqueFileNameDedup
 ```
  should be aware of using this function can hide some "bad design" installer project.

Again, in scenario described by @apawsey, dedupe is only correct if his "excelPluginFeature" is a child feature of "clientFeature". Otherwise, one feature will have the file missing if installed alone.

**Comment by oleg-shilo** [comment]
Agree. The use-case for this feature is not very common. Thus the code API documentation provides the link tho this very discussion:

![image](https://user-images.githubusercontent.com/16729806/37283883-b7682ac6-264d-11e8-837e-38ace2ff2933.png)

**Comment by Lumoryel** [comment]
Hi there, first of all big compliments and many thanks for this project!

I tried "project.WildCardDedup = Project.UniqueFileNameDedup" but it does not do anything.
When I do it manually, it works.
What am I doing wrong?
Thanks in advance
Henning

**Comment by oleg-shilo** [comment]
> Hi there, first of all big compliments and many thanks for this project!

Not a problem, glad you like it.

> What am I doing wrong?
I am not sure what you mean "manually", but I suggest you just debug it.

The default `dedup` implementation is very simple and you can easily repeat it in your codebase so you can put a breakpoint (or `Debug.Assert(false);`) there and see what is wrong with it:
```C#
        public static Action<Dir> UniqueFileNameDedup = dir =>
        {
            // Issue #270: Deduplication of files added with wildcards sample dedup
            if (dir.Files.Any())
                dir.Files = dir.Files.DistinctBy(x => x.Name.PathGetFileName()).ToArray();
        };
```

**Comment by Lumoryel** [comment]
With "manually" I mean I write the code on my own:

```
project.ResolveWildCards();

foreach (Dir projectDir in project.Dirs)
{
    DedupeFolder(projectDir);
}

void DedupeFolder(Dir dir)
{
    if (dir.Files.Any())
        dir.Files = dir.Files.DistinctBy(x => Path.GetFileName(x.Name)).ToArray();
    foreach (Dir childDir in dir.Dirs)
    {
        DedupeFolder(childDir);
    }
}


```

If I instead call "project.WildCardDedup = Project.UniqueFileNameDedup", then nothing happens. Like the project never uses "WildCardDedup".

**Comment by Lumoryel** [comment]
Directly after "var project = new Project();"

**Comment by oleg-shilo** [comment]
never mind, I found the probem. Something is broken in `ResolveWildCards`.
So it is a defect. Reopening the issue.
It should not be too complicated. Will issue the hotfix asap.

**Comment by oleg-shilo** [comment]
Found it. At least in my experiment.
I used `DirFile(@"a\b\*.*")`
Dedup is called on a freshly discovered folder. But DirFile does not discover folders but files only. Thus one can pass a file filter directly in the `DirFile` signature.
And the directories are discovered by `Files("*.*")` as it goes recursively into the all subfolders (discovered folders).

I need to make it more clear in the  `WildCardDedup` documentation that this delegate is only invoked if you use `Files`.
I will consider a global dedup that is automatically applied on ALL dirs regardless if they are specified by user or discovered. Maybe even instead of the current approach as it clearly confuses the users.

As for you you can easily mimic global dedup with this:
```C#
...
project.ResolveWildCards();
project.AllDirs.ForEach(Project.UniqueFileNameDedup);
```

Note, I used `project.AllDirs`. `project.Dirs` only returns the top level dirs but not their children.

**Comment by oleg-shilo** [comment]
Please try `https://www.nuget.org/packages/WixSharp`

I decided not to resort to global deduplication as it is a little counter-intuitive. If a dir/file is added by the user then wixSharp has no business messing with what the user added. It is for him/her to fix after compiling error is reported.

However, if it is WixSharp who added/aggregated the dir/file then it is OK for it to try to fix the collision in the added WixEntities according to user-defined algorithm.

So what is new in this hot-fix (v1.20.4.0)?
The dedup routine was invoked for all discovered directories. Now it is invoked even for the user-defined directories if they have WixSharp-added files (`DirFiles`).

PLease let me know if it fixed the problem

**Comment by Lumoryel** [comment]
Works like charm now! Thank you Oleg :)

**Comment by oleg-shilo** [comment]
Great, thank you for letting me know

**Comment by Lumoryel** [comment]
I recognized that the dedup method is called multiple times for the same directory, even though I only added wildcard files for that directory only once. Is there any explanation ?
It is also called multiple times if I to multiple "adds" into the same target directory, well I could expect that but it would not be necessary?

**Comment by oleg-shilo** [comment]
Checking....

**Comment by oleg-shilo** [comment]
Sorry, it took much longer to respond. I needed to fulfil my commitments to other OSS projects.

> I recognized that the dedup method is called multiple times for the same directory, even though I only added wildcard files for that directory only once. Is there any explanation ?

I have created a dedicated test that shows that the dedup is called only once for every discovered directory. So, not sure that what you are describing takes place.

> It is also called multiple times if I to multiple "adds" into the same target directory...

What are you referring to when you say "to multiple adds". Is it covered by the unit test I shared above?

**Comment by Lumoryel** [comment]
```
        var p = new ManagedProject("TestProduct");
        p.WildCardDedup = dir =>
        {
            System.IO.File.AppendAllText(@"D:\Temp\DedupCount.txt", "Dedup called\n");
            if (dir.Files.Any())
            {
                dir.Files = dir.Files.DistinctBy((File x) => x.Name.PathGetFileName()).ToArray();
            }
        };
        p.AddDir(new Dir("TestDir", 
            new Files(@"D:\Temp\WIXTestFolder1\*.txt"), 
            new Files(@"D:\Temp\WIXTestFolder2\*.txt") // containing same two .txt files as WIXTestFolder1
            ));
        p.ResolveWildCards();
```

When I do this, the dedup is called twice although I have only one target Dir that contains duplicate files.

**Comment by oleg-shilo** [comment]
I think there is still a misunderstanding about the purpose of WildCardDedup. 

> ...only one target Dir that contains duplicate files.
Target dir does not matter. 

I have created another test for your ver scenario above:
```C#
[Fact]
[Description("Issue #270 (WildCardDedup2)")]
public void Fix_Issue_270_WildCardDedup_for_discovered_dirs2()
{
    var file1 = Environment.CurrentDirectory.PathJoin("dummydir3", "test.txt");
    var file2 = Environment.CurrentDirectory.PathJoin("dummydir4", "test.txt");

    System.IO.Directory.CreateDirectory(file1.PathGetDirName());
    System.IO.Directory.CreateDirectory(file2.PathGetDirName());

    int dedupInvokeCounter = 0;

    var project = new Project("MyProduct");

    project.AddDir(new Dir("TestDir",
                new Files(file1.PathGetDirName().PathJoin(@"*.txt")),
                new Files(file1.PathGetDirName().PathJoin(@"*.txt"))));

    project.WildCardDedup = dir =>
    {
        dedupInvokeCounter++;
    };

    project.ResolveWildCards();

    // '0' because there are no sub-folders in the dummydir3 and dummydir3 dirs
    // WildCardDedup is only used to handle duplications in the source folders discovered by `Files(...)`
    Assert.Equal(0, dedupInvokeCounter);
}
```

Project dudup is provided to handle source dirs that are not added by the user but by the compiler as the result of resolving wild card in `Files`, not the target dirs and files. Look at these scenarios:

---

One target dir "TestDir"
Two source files "c:\temp\a\test.txt" and "c:\temp\d\test.txt"
The user is responsible for handling the duplication.

```c#
project.AddDir(
    new Dir("TestDir",
        new File(@"c:\a\test.txt"),
        new File(@"c:\b\test.txt")),
```

---

One target dir "TestDir"
Multiple source files present in "c:\a" and "c:\d"
The user is responsible for handling possible duplication. 
```c#
project.AddDir(
    new Dir("TestDir",
        new DirFiles(@"c:\a"),
        new DirFiles(@"c:\b")),
```

---

One target dir "TestDir"
Multiple source files present in "c:\a" and "c:\b"

```C#
If "c:\a" and "c:\b" have subfolders // (e.g. "c" and "d") then two new source dirs sill be detected: 
    Discovered two new source directories "c:\a\c" and "c:\d\d"
    Added multiple source files present in "c:\a\c" and "c:\b\d"
    // WixSharp is responsible for handling unknown duplications in "c:\a\c" and "c:\d\d". 
    project.WildCardDedup will be called twice (once per discovered directory). 
else if c:\a" and "c:\b" have no subfolders
    project.WildCardDedup will not be invoked 
```

The user is responsible for handling possible duplication in the rest of the cases. 
```c#
project.AddDir(
    new Dir("TestDir",
        new Files(@"c:\a\*txt"),
        new Files(@"c:\b\*.txt")),
```

The `project.WildCardDedup` explains this better now.

**Comment by Lumoryel** [comment]
Thanks Oleg, but I still don't get it.

You say in your example with "new Files(*.txt)":

> "else if c:\a" and "c:\b" have no subfolders project.WildCardDedup will not be invoked"

I do not have subfolders but WildCardDedup is called twice!


Another fact why I thought the target dir is deduplicated:

When I try to build the msi with my example
```
        p.AddDir(new Dir("TestDir",
            new Files(@"D:\Temp\WIXTestFolder1\*.txt"),
            new Files(@"D:\Temp\WIXTestFolder2\*.txt") // containing same .txt file as WIXTestFolder1
            ));
```

WITHOUT defining the WildCardDedup, then I get errors like

` error LGHT0204: ICE30: The target file 'Test1.txt' is installed in '[TARGETDIR]\TestDir\' by two different components on an SFN system: 'Component.Test1.txt_793771408' and 'Component.Test1.txt_856227216'. This breaks component reference counting.`

But WITH WildCardDedup, there are no errors. Why is this, if you say the wildcarddedup does only dedup **subfolders** of wildcard adds?

By the way when I do

```
        p.AddDir(new Dir("TestDir",
            new DirFiles(@"D:\Temp\WIXTestFolder1"),
            new DirFiles(@"D:\Temp\WIXTestFolder2")
            ));
```

then there is no msi build error, even if I don't use WildCardDedup.
I don't understand why "new Files(*.*)" will produce an error whilst new DirFiles() won't?

Sorry for my misunderstandings and thanks for explaining this for me!


Why I generally do this in our setup:
I want to add multiple dll projects to the same target directory AND get all their dependencies (nugets etc.) automatically, so I'm using Files(*.dll) in those dll project folders.
Of course there are duplicate files like nugets or common own dlls, but when I ensure that they have the same version, then it should work IMHO and this saves a lot of wixsharp maintenance (adding/adjusting all those dependencies manually).
With your WildCardDedup, I no longer get errors when building the msi, and it's not that it only deduplicates subfolders so I'm still missing somthing.

**Comment by oleg-shilo** [comment]
OK, it looks like we are dealing with different behaviour and I am not looking at the test case that shows the behaviour you are observing.
Let's do this, can you send me a hello world style sample that shows the problem. Not the code snippet but a complete VS project and I will investigate.

**Comment by oleg-shilo** [comment]
I have just created a fresh project from one of the WixSharp templates. Replaced Main with your code sample from https://github.com/oleg-shilo/wixsharp/issues/270#issuecomment-1611279724.

After building it it did not call dedup, as expected.

I have attached the zip file with all code, binaries and even source dirs for `Files` routine.

Have a look and let me know if it behaves the same way for you.
[WixSharp Setup9.zip](https://github.com/oleg-shilo/wixsharp/files/11913881/WixSharp.Setup9.zip)

**Comment by Lumoryel** [comment]
I think you have a copy paste error:

```
                new Files(@".\WIXTestFolder1\*.txt"),
                new Files(@".\WIXTestFolder1\*.txt") // containing same two .txt files as WIXTestFolder1
```

The second one should be WIXTestFolder2

I recognized that it does not happen with WixSharp nuget version 1.19 (as you used in your project), but it happens with 1.20.4.

**Comment by oleg-shilo** [comment]
Great, after upgrading the package I got the error reproduced.

That's why I always ask for a test project for the difficult to reproduce cases.

**Comment by oleg-shilo** [comment]
First of all, I want to thank you for being so persistent when trying to get to the bottom of this. It not only helped me to find the cause of observing different behaviour in our experiments but also reevaluate the usability of the whole dedup approach.
  
**The cause of our very  problem**
Because of the merging mistake in the the `wix-v3-master` branch, it was missing a single line that was the reason for our discrepancies.

**The further considerations**
After experiencing the difficulties with clear articualting the dedup intent to the user(s) and challenges with troubleshooting the feature I decided to reevaluate current usability of the feature and basically reinvent (and re-implement) it with this very clear definition:

> When `ResolveWildCards` is invoked, it inserts the files discovered from project's `DirFiles` and `Files` definitions. For every `Dir`, which getst new discovered files, the global dedup delegate `Project.WildCardDedup` will be invoked.

Thus there is no any more difference between `DirFiles` and `Files`. There is no difference between user defined and discovered `Dir`s. Thus the dedup becomes truly global.

It does not make sense to continue the discussion here so I will create anew issue for fixing `WildCardDedup` usability problems.
